core/src/main/java/org/apache/druid/data/input/impl/TimedShutoffInputSourceReader.java

```diff
@@ -131,6 +134,7 @@ public void close() throws IOException
         shutoffTime.getMillis() - System.currentTimeMillis(),
         TimeUnit.MILLISECONDS
     );
+    exec.shutdown();
```

Comment:
This is necessary to shutdown the `exec` created in this class. Previously it was being shutdown via the registering with the Closer. However, that is bad form since the Closer can be executed by a thread in this `exec`. I removed the shutdown of this `exec` from the Closer and added this line. This means that the scheduled task (which calls the iterator.close() method will always run), which is ok since we have the atomic boolean to ensure that the iterator.close() is idempotent. After the task ran then the `exec` will just automatically close. This no longer relies on the Closer, the thread in the `exec` or the iterator.close().

Reply:
I see, so its purpose is to make sure the executor thread is stopped after the task fires. OK, that makes sense.